### DISCUSSION TITLE
Circuit with random start and how to avoid ?

### DISCUSSION THREAD

**Discussion by jongena** [hneemann/Digital#1063] [Q&A] [answered]
I have the following circuit: a clock (with two phase output ΦC and ΦT ) and two connected D-flip-flops using each as clock input the 
inversion of one of the phases (see attached) 
The ouputs of the second D-flip flop look random at the start of the circuit. Sometimes it starts with output 2 high and output 3 low
and sometimes it is the reverse. Output 1 on the other hand is always the same. See attached output files.

I also constructed a 16 bit shift register, with 16 consecutive D flip flops (again with inverted C input) showing similar effects.
Another side effect seems to be that the D input is immediately visible at the output, thus overwriting the default output.

Further tests showed that when the inversion is removed at the C input, the random behaviour disappeared, and also the default output then appeared on the Q output. 

My question is how this effect with the inverted C input should be understood, and equally important, what can be done to assure that only one of the two options will be output at the start of the simulation when the inverted C input is in use?

Shift registers of various lengths with consecutive D flip flops are used in the circuit I have at hand and determine the state of other ones. The seemingly random starting conditions (as seen in one of the photos with the traces) set the whole circuit in a random state at the beginning of the simulation.

From what I have seen in Digital over the past months, it is a very powerful tool, so I am sure that there must be a way to tackle this. 
My questions:
1. How to make sure that always only one of the two possible initial states will appear ? Is there a way to force the Digital simulation to always start in exactly the same way ?
2. Is there a perhaps better way to implement a shift register than  with  consecutive D flip-flops ?

Am I  making an elephant out of a fly ...  ? Perhaps... but as a novice in this field I am still learning every day !
Any hint is very welcome. 

![IMG_3499](https://user-images.githubusercontent.com/108168531/204020750-d35314e4-d4fd-4242-980b-bf13b2650699.jpg)
![IMG_3500](https://user-images.githubusercontent.com/108168531/204020772-c628d9af-44c6-4b61-9061-319ce3f9ea82.jpg)
<img width="1607" alt="Simple circuit with random start" src="https://user-images.githubusercontent.com/108168531/204021204-cfb6f302-0f24-458b-94cc-a09cccd18852.png">

**Comment by hneemann** [comment, marked as answer]
You can use the reset component to ensure that the circuit always starts in the same way. 
In general, it is bad practice to generate the clock signal of a component by logic gates. 
It is better to use a common clock signal for all flip-flops.

**Reply by jongena** [reply]
Thanks for the reply. I fully agree that using "derived" clocks is not the best option. But it is not my choice... the circuit I am trying 
to simulate (for a detailed understanding) is that from a 1971 Japanese calculator and the design is like that... 
How best to insert the reset component in the bottom circuit with the two D flip-flops ? Or should I construct the D flip flop with gates and insert the reset component in that circuit ? Can you provide me with an example ?
Thanks in advance !

**Reply by hneemann** [reply]
One possibility would be to replace a D flip-flop with a register. This has an enable input and this can be held low during the switch-on phase with the reset component. What works best in your circuit is hard to say. Can you provide the circuit?

**Reply by hneemann** [reply]
The problem arises because the flip-flops are operated with inverted clock, but the clock signal has the default value zero. At the moment of power-on all flip-flops see a rising edge at the clock. This is processed in random order, which leads to the problem.
If the default value of the clock input is set to one, the problem disappears.

**Reply by jongena** [reply]
Thank you for the suggestions ! I have implemented the default value "one" in the clock inputs of the various sub circuits.
It helps but the problem is not fully gone. To illustrate this, I have attached the simulation of the keyboard input. To this end, I have attached all necessary files and custom components and added .txt to the filename extension to be able to attach them. Github seems not to accept .dig files.  Simply removing .txt at the end of the filename is all that is required to make them ready to run.

Files needed to run that part of the simulation are: 

Keyboard_interface_problem.dig.txt
Digit-phases-problem.dig.txt
Bit-phases-problem.dig.txt
Two-phase-clock-problem.dig.txt
D flip-flop-with-set-problem.dig.txt
D-flip-flop-problem-positive-D.dig.txt
M5812-problem.dig.txt


A few words of explanation:
The old japanese calculator that I am simulating is a state machine with 15 states, operating with negative logic. This old calculator uses not one clock signal, but a two phase clock: one phase captures the input, the other phase causes the transfer from input to output. The components in the calculator are triggered by falling edges in the two phases of the clock. Thus I built components using  the standard D-flip-flops in Digital as building blocks, as you have already seen and also present in the attached files. Note that I also had to synchronize the two phase clock signals, the bit and digit phases. Not doing so led to unwanted addtional blips in various signals. This is a consequence of deriving clock signals in a chain, as you already remarked earlier. 

I also inserted a reset component in the RS flip-flops that are also present (not using the standard components in Digital, but implemented using "crossed gates";see the OPxx~ flags at the right in the circuit). That seems to work well.

At other places, and it seems always to be the case when using D flip-flops, random outputs still appear.  
The attached example illustrates clearly the various questions I have.  
See e.g. when running the file keyboard-interface-problem, and looking at the following: 
- the flags KPN~ and KPN that still behave randomly at the start of the simulation (although in all custom D-flip-flops, the clock input has been set to "one" by default). Why does the KPN output behave randomly and not the outputs KPR and KPD, the only difference being the combination of gates that lead up to KPN ? 
- the various bit phases ΦBxE~ : they are randomly positive and negative at start. 
- the flags OPE~ and OPE 
- the two custom D-flip-flops that produce delayKASCE~; why does this not happen at the delayKD~ and delayKM~ versions ? 
- the first custom D-flip-flop in the series of 5 that produces the POC~ output

The question is more general:
Under which circumstances can one expect a random output  ?  Is there a way I can check /predict this ?
But also: 
Perhaps I set the defaults in the wrong place ? 
Perhaps I need anyway to make use of a register in combination with the reset component here and there ? But how 
to determine in advance where this would be needed ? 

Any advice is very welcome !

With my best regards,
J.Ongena Belgium

[Keyboard_interface_problem.dig.txt](https://github.com/hneemann/Digital/files/10160299/Keyboard_interface_problem.dig.txt)
[Bit-phases-problem.dig.txt](https://github.com/hneemann/Digital/files/10160300/Bit-phases-problem.dig.txt)
[M5812-problem.dig.txt](https://github.com/hneemann/Digital/files/10160301/M5812-problem.dig.txt)
[Digit-phases-problem.dig.txt](https://github.com/hneemann/Digital/files/10160302/Digit-phases-problem.dig.txt)
[Two-phase-clock-problem.dig.txt](https://github.com/hneemann/Digital/files/10160303/Two-phase-clock-problem.dig.txt)
[D flip-flop-with-set-problem.dig.txt](https://github.com/hneemann/Digital/files/10160304/D.flip-flop-with-set-problem.dig.txt)
[D-flip-flop-problem-positive-D.dig.txt](https://github.com/hneemann/Digital/files/10160305/D-flip-flop-problem-positive-D.dig.txt)

**Reply by jongena** [reply]
I continued my tests. I keep having random output using the standard D flip-flops of Digital. So I replaced all "two phase clock" D flip-flops using registers and the enabling option with the reset component. This leads indeed to a standard start, which is excellent !  The only problem that remains is that I need to set the output of these custom "two phase clock" D flip-flips to default high (because of the negative logic). I get otherwise conflicting conditions at the start of the simulation. From what I understand this seems to come down to set the default output of the register to high. However, I do not see a direct possibility to do so. 

Q1:  Is there a way to enforce a positive Q output as a default for the register (or perhaps, more in general, enforcing default outputs to custom components)  ? 
Q2: the best solution still (because of additional delays that otherwise need to be corrected throughout the simulation) seems to be able to use the standard D flip-flop of DIgital, but then such that it always start in the same way. Any possibility to do this with the reset component ?
Q3: I also built  D flip-flops as custom components ("using crossed gates"), with the reset component inside to define "Q = high" as the default output. But these custom components do not behave exactly as the standard D flip flop of Digital: the Q and ~Q outputs have a small phase where they are at the same logical level; with the standard D flip-flop the Q and ~Q outputs are always exactly opposite to each other. This small phase where Q and ~Q are at the same level leads to "oscillation errors" in other parts of the simulation. 

How to proceed here without having to revise the whole simulation...  (thereby avoiding all kinds of other difficulties...) ? 
To my opinion, the best would be to use the standard D- flip flops in Digital, but with the possibility to use the reset component, to avoid random outputs at the start. 

Of course, any other suggestion is welcome.... to arrive at both (i) a stable start of the simulation using the reset component (or by any other means) and (ii) with "high" as a default for the Q output of the D-flip-flop, register, any other (custom) component that could be suitable, or an additional circuit that helps to realize this double requirement. 

Any advice is highly welcome.

**Reply by jongena** [reply]
Continuing the investigations, here is what I think is the best solution for the D flip flop that I need: using an asynchronous D flip flop combined with the reset component, as per the circuit below. I implemented it, and seems to realize the two conditions above: I now get a unique starting point at the start of the simulation, with the Q output at high level. 
Let me know if this could be realized even simpler  ?


<img width="1070" alt="Screenshot 2022-12-13 at 20 12 56" src="https://user-images.githubusercontent.com/108168531/207424588-d8a94a46-6de2-4bda-bb89-c1e9325cf071.png">

**Reply by hneemann** [reply]
@jongena 
That's a pretty good idea! I'm surprised I didn't come up with it myself, because that's exactly how you would build it with real hardware. Especially considering that a similar implementation has already been proposed here at some point, but I can't find the old issue:

![D-T](https://user-images.githubusercontent.com/7190563/207533291-d22afa8f-943b-4c01-8373-1786381972bc.png)

Using the asynchronous flip-flop makes the circuit much simpler, of course.

**Reply by jnoche** [reply]
"_I can't find the old issue._"  Are you referring to https://github.com/hneemann/Digital/discussions/1028 ?

**Reply by hneemann** [reply]
@jnoche 
Yes! Thank you!
